comms/src/protocol/messaging/messaging.rs

```diff
@@ -0,0 +1,270 @@
+// 2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the
+// following disclaimer in the documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote
+// products derived from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+// INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+// USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+use super::error::MessagingProtocolError;
+use crate::{
+    compat::IoCompat,
+    connection_manager::next::ConnectionManagerRequester,
+    message::InboundMessage,
+    outbound_message_service::{MessageTag, OutboundMessage},
+    peer_manager::{AsyncPeerManager, NodeId, NodeIdentity},
+    protocol::{
+        messaging::{inbound::InboundMessaging, outbound::OutboundMessaging},
+        ProtocolEvent,
+        ProtocolNotification,
+    },
+    types::CommsSubstream,
+};
+use bytes::Bytes;
+use futures::{
+    channel::{mpsc, oneshot},
+    stream::Fuse,
+    AsyncRead,
+    AsyncWrite,
+    SinkExt,
+    StreamExt,
+};
+use log::*;
+use std::{
+    collections::{hash_map::Entry, HashMap},
+    sync::Arc,
+    time::Duration,
+};
+use tokio::{runtime, time::delay_for};
+use tokio_util::codec::{Framed, LengthDelimitedCodec};
+
+const LOG_TARGET: &str = "comms::protocol::messaging";
+pub const PROTOCOL_MESSAGING: Bytes = Bytes::from_static(b"/tari/messaging/0.1.0");
+const MAX_SEND_MSG_ATTEMPTS: usize = 4;
+/// The size of the buffered channel used for _each_ peer's message queue
+const MESSAGE_QUEUE_BUF_SIZE: usize = 20;
+
+/// Request types for MessagingProtocol
+pub enum MessagingRequest {
+    SendMessage(OutboundMessage, oneshot::Sender<Result<(), MessagingProtocolError>>),
+}
+
+pub enum MessagingEvent {
+    MessageReceived(Box<NodeId>, InboundMessage),
+    InvalidMessageReceived(Box<NodeId>),
+    SendMessageFailed(OutboundMessage),
+    SendMessageSucceeded(MessageTag),
+}
+
+pub struct Messaging {
+    executor: runtime::Handle,
+    connection_manager_requester: ConnectionManagerRequester,
+    node_identity: Arc<NodeIdentity>,
+    peer_manager: AsyncPeerManager,
+    proto_notification: Fuse<mpsc::Receiver<ProtocolNotification<CommsSubstream>>>,
```

Comment:
Readability nit: proto means protobuf everywhere else I think this should be `protocol_notification`

Reply:
Sort of agree - but this is a private field and the type should quickly clear up what this field is referring to (I think clearly nothing to do with protobuf) - however, I'll change it to ensure that it's clear

Reply:
Really not a big deal but I have to say that working with the proto files can be complicated and tricky. The go to definition doesn't always work and for the future people I really feel we should do as much as possible to make it easy to parse.

Reply:
This is an example of 3 extra characters means you don't have to rely on context anymore by being explicit.